pkg/program/shell/run_dir_unix_test.go

```diff
@@ -0,0 +1,102 @@
+// +build !windows,!plan9
+
+package shell
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/elves/elvish/pkg/util"
+)
+
+// RunDirPathTest is a test case to be used in getRunDirPath.
+type RunDirPathTest struct {
+	Name          string
+	XDGRuntimeDir string
+	Tmpdir        string
+	Want          string
+}
+
+func TestGetSecureRunDir_UseXDGWhenPresent(t *testing.T) {
+	dst, cleanup := util.TestDir()
+	defer cleanup()
+
+	envOverrides := map[string]string{"XDG_RUNTIME_DIR": dst}
+
+	withEnvVarOverrides(envOverrides, func() {
+		runDir, err := getSecureRunDir()
+
+		if err != nil {
+			t.Errorf("Could not get run dir: %v", err)
+		}
+
+		wantRunDir := filepath.Join(dst, "elvish")
+		if runDir != wantRunDir {
+			t.Errorf("Got run dir %v, want %v", runDir, wantRunDir)
+		}
+	})
+}
+
+func TestGetRunDirPath(t *testing.T) {
+	tests := []RunDirPathTest{
+		{
+			Name:          "use XDG_RUNTIME_DIR when present",
+			XDGRuntimeDir: "/foo",
+			Tmpdir:        "",
+			Want:          "/foo/elvish",
+		},
+		{
+			Name:          "fallback to TMPDIR",
+			XDGRuntimeDir: "",
+			Tmpdir:        "/foo",
+			Want:          fmt.Sprintf("/foo/elvish-%d", os.Getuid()),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.Name, func(t *testing.T) {
+			t.Helper()
+
+			envOverrides := map[string]string{
+				"XDG_RUNTIME_DIR": test.XDGRuntimeDir,
+				"TMPDIR":          test.Tmpdir,
+			}
+
+			withEnvVarOverrides(envOverrides, func() {
+				runDir := getRunDirPath()
+				wantRunDir := test.Want
+
+				if runDir != wantRunDir {
+					t.Errorf("Got run dir %v, want %v", runDir, wantRunDir)
+				}
+			})
+		})
+	}
+}
+
+func withEnvVarOverrides(envOverrides map[string]string, fn func()) {
```

Comment:
This function is better written with the following signature:

```go
func withEnvVarOverrides(overrides map[string]string) func()
```

And used like this:

```go
restore := withTempOverrides(...)
defer restore()
```

In fact, this is exactly the style used by `util.TestDir` you are using in one of the tests. It suits the style of Go much better.

Also, there is already a testing helper, [`util.TempEnv`](https://github.com/elves/elvish/blob/7325c89be4131a8fc22e4d405f474f31ef56f11b/pkg/util/temp_env.go#L5), that does something very similar, except that it only supports one environment variable, and treats non-existing variables as if they are empty.

I'd move this function to the `util` package, call it `WithTempEnvs`, and implement `WithTempEnv` as just a thin wrapper around it. But that means you also need to have unit tests against `WithTempEnvs` too, which will make the PR bigger and slower to merge, so I'm perfectly happy if you act on my suggestion for the signature, and keep this helper in this test file for now.